src/V3Error.h

```diff
@@ -276,33 +283,41 @@ class V3Error final {
     // CONSTRUCTORS
     // ACCESSORS
     static void debugDefault(int level) { s_debugDefault = level; }
-    static int debugDefault() { return s_debugDefault; }
+    static int debugDefault() VL_MT_SAFE { return s_debugDefault; }
     static void errorLimit(int level) { s_errorLimit = level; }
-    static int errorLimit() { return s_errorLimit; }
+    static int errorLimit() VL_MT_SAFE { return s_errorLimit; }
     static void warnFatal(bool flag) { s_warnFatal = flag; }
-    static bool warnFatal() { return s_warnFatal; }
-    static string msgPrefix();  // returns %Error/%Warn
-    static int errorCount() { return s_errCount; }
-    static int warnCount() { return s_warnCount; }
-    static bool errorContexted() { return s_errorContexted; }
+    static bool warnFatal() VL_MT_SAFE { return s_warnFatal; }
+    static string msgPrefix() VL_MT_SAFE;  // returns %Error/%Warn
+    static string msgPrefixNoLock() VL_REQUIRES(s_mutex) VL_MT_SAFE;  // returns %Error/%Warn
+    static int errorCount() VL_MT_SAFE { return s_errCount; }
+    static int warnCount() VL_MT_SAFE { return s_warnCount; }
+    static bool errorContexted() VL_MT_SAFE { return s_errorContexted; }
     static void errorContexted(bool flag) { s_errorContexted = flag; }
     // METHODS
-    static void incErrors();
+    static void incErrors() VL_MT_SAFE;
     static void incWarnings() { s_warnCount++; }
-    static void init();
+    static void init() VL_MT_SAFE;
     static void abortIfErrors() {
         if (errorCount()) abortIfWarnings();
     }
     static void abortIfWarnings();
-    static void suppressThisWarning();  // Suppress next %Warn if user has it off
-    static void pretendError(V3ErrorCode code, bool flag) { s_pretendError[code] = flag; }
-    static bool isError(V3ErrorCode code, bool supp);
+    static void suppressThisWarning() VL_MT_SAFE;  // Suppress next %Warn if user has it off
+    static void pretendError(V3ErrorCode code, bool flag) VL_MT_SAFE {
+        const VerilatedLockGuard lg{s_mutex};
+        s_pretendError[code] = flag;
+    }
+    static bool isError(V3ErrorCode code, bool supp) VL_REQUIRES(s_mutex) VL_MT_SAFE;
     static string lineStr(const char* filename, int lineno);
-    static V3ErrorCode errorCode() { return s_errorCode; }
-    static void errorExitCb(ErrorExitCb cb) { s_errorExitCb = cb; }
+    static V3ErrorCode errorCode() {
+        const VerilatedLockGuard lg{s_mutex};
+        return s_errorCode;
+    }
+    static void errorExitCb(ErrorExitCb cb) VL_REQUIRES(s_mutex) { s_errorExitCb = cb; }
 
     // When printing an error/warning, print prefix for multiline message
-    static string warnMore();
+    static string warnMore() VL_MT_SAFE;
+    static string warnMoreNoLock() VL_REQUIRES(s_mutex) VL_MT_SAFE;
```

Comment:
I think the mix of lock and no-lock functions is hard to maintain.  I suggest instead make a separate class with all protected or private members, which assumes that the lock has been obtained and those functions are only called from that class, or from this class which first obtains the lock.